spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java

```diff
@@ -0,0 +1,197 @@
+import org.apache.spark.sql.Column;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Encoders;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.util.SerializableConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import parquet.Preconditions;
+
+public class RemoveOrphanFilesAction implements Action<RemoveOrphanFilesActionResult> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RemoveOrphanFilesAction.class);
+
+  private final SparkSession spark;
+  private final JavaSparkContext sparkContext;
+  private final SerializableConfiguration hadoopConf;
+  private final FileIO fileIO;
+  private final int partitionDiscoveryParallelism;
+  private final String dataLocation;
+
+  private String allDataFilesTable = null;
+  private Long olderThanTimestamp = null;
+  private boolean dryRun = false;
+
+  RemoveOrphanFilesAction(SparkSession spark, Table table) {
+    this.spark = spark;
+    this.sparkContext = new JavaSparkContext(spark.sparkContext());
+    this.hadoopConf = new SerializableConfiguration(spark.sessionState().newHadoopConf());
+    this.fileIO = table.io();
+    this.partitionDiscoveryParallelism = spark.sessionState().conf().parallelPartitionDiscoveryParallelism();
+    this.dataLocation = table.locationProvider().dataLocation();
+  }
+
+  public RemoveOrphanFilesAction allDataFilesTable(String newAllDataFilesTable) {
+    this.allDataFilesTable = newAllDataFilesTable;
+    return this;
+  }
+
+  public RemoveOrphanFilesAction olderThan(long newOlderThanTimestamp) {
+    this.olderThanTimestamp = newOlderThanTimestamp;
+    return this;
+  }
+
+  public RemoveOrphanFilesAction dryRun(boolean newDryRun) {
+    this.dryRun = newDryRun;
+    return this;
+  }
+
+  @Override
+  public RemoveOrphanFilesActionResult execute() {
+    Preconditions.checkArgument(allDataFilesTable != null, "allDataFilesTable must be set");
+    Preconditions.checkArgument(olderThanTimestamp != null, "olderThanTimestamp should be set");
+
+    Dataset<Row> validDataFileDF = buildValidDataFileDF();
+    Dataset<Row> actualDataFileDF = buildActualDataFileDF();
+
+    Column joinCond = validDataFileDF.col("file_path").equalTo(actualDataFileDF.col("file_path"));
+    List<String> orphanDataFiles = actualDataFileDF.join(validDataFileDF, joinCond, "leftanti")
+        .as(Encoders.STRING())
+        .collectAsList();
+
+    if (!dryRun) {
+      Tasks.foreach(orphanDataFiles)
+          .noRetry()
+          .suppressFailureWhenFinished()
+          .onFailure((file, exc) -> LOG.warn("Failed to delete data file: {}", file, exc))
+          .run(fileIO::deleteFile);
+    }
+
+    return new RemoveOrphanFilesActionResult(orphanDataFiles);
+  }
+
+  private Dataset<Row> buildValidDataFileDF() {
+    return spark.read().format("iceberg")
+        .load(allDataFilesTable)
+        .select("file_path")
+        .distinct();
```

Comment:
You probably don't need `distinct` if you're using this in a left anti-join. That just introduces an additional shuffle.

Reply:
Done.